From 003de52b35add6a9ff157d06b0f9ffda8586df34 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Fri, 6 Sep 2019 23:54:27 +0900
Subject: [PATCH] Finish memory leak

Previously BN_MONT_CTX objects were not freed.  The default RSA method
uses static function rsa_ossl_finish() to free them.  This fix gets
the reference to the function and call it from our finish function.
---
 neverbleed.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/neverbleed.c b/neverbleed.c
index 420665e..552ddd9 100644
--- a/neverbleed.c
+++ b/neverbleed.c
@@ -1206,12 +1206,17 @@ Redo:
     _exit(0);
 }
 
+static int (*rsa_finish)(RSA *rsa);
+
 static int priv_rsa_finish(RSA *rsa)
 {
     struct st_neverbleed_rsa_exdata_t *exdata;
     struct st_neverbleed_thread_data_t *thdata;
 
     get_privsep_data(rsa, &exdata, &thdata);
+
+    rsa_finish(rsa);
+
     if (exdata == NULL)
         return 1;
 
@@ -1406,6 +1411,8 @@ int neverbleed_init(neverbleed_t *nb, char *errbuf)
     const EC_KEY_METHOD *ecdsa_default_method;
     RSA_METHOD *rsa_method = RSA_meth_dup(RSA_PKCS1_OpenSSL());
 
+    rsa_finish = RSA_meth_get_finish(rsa_method);
+
     RSA_meth_set1_name(rsa_method, "privsep RSA method");
     RSA_meth_set_priv_enc(rsa_method, priv_enc_proxy);
     RSA_meth_set_priv_dec(rsa_method, priv_dec_proxy);
-- 
2.17.1

